rotkehlchen/exchanges/coinbaseprime.py

```diff
@@ -0,0 +1,477 @@
+        try:
+            response = self.session.get(url=uri, params=params)
+        except requests.RequestException as e:
+            raise RemoteError(f'Coinbase Prime API request failed due to {e}') from e
+
+        try:
+            data = response.json()
+        except requests.JSONDecodeError as e:
+            raise RemoteError(f'Coinbase Prime returned invalid json {response.text}') from e
+
+        return data
+
+    @overload
+    def _query_paginated_endpoint(
+            self,
+            query_params: dict[str, Any],
+            portfolio_id: str,
+            method: Literal['transactions'],
+            decoding_logic: Callable[[dict[str, Any]], AssetMovement | None],
+    ) -> list[AssetMovement]:
+        ...
+
+    @overload
+    def _query_paginated_endpoint(
+            self,
+            query_params: dict[str, Any],
+            portfolio_id: str,
+            method: Literal['orders'],
+            decoding_logic: Callable[[dict[str, Any]], Trade | None],
+    ) -> list[Trade]:
+        ...
+
+    def _query_paginated_endpoint(
+            self,
+            query_params: dict[str, Any],
+            portfolio_id: str,
+            method: Literal['orders', 'transactions'],
+            decoding_logic: Callable[[dict[str, Any]], AssetMovement | None] | Callable[[dict[str, Any]], Trade | None],  # noqa: E501
+    ) -> list[Trade] | list[AssetMovement]:
+        """Abstraction to consume all the events in the selected queries.
+        It uses the `decoding_logic` to process the different events and returns a list
+        whose contents depend on the function's called arguments.
+
+        This function may raise:
+        - RemoteError
+        """
+        result = []
+        while True:
+            response = self._api_query(
+                module='portfolios',
+                path=f'{portfolio_id}/{method}',
+                params=query_params,
+            )
+
+            for raw_event in response[method]:
+                try:
+                    if (event := decoding_logic(raw_event)) is None:
+                        log.warning(
+                            f'Wont process event {raw_event} from coinbase prime. Skipping',
+                        )
+                    else:
+                        result.append(event)
+                except DeserializationError as e:
+                    self.msg_aggregator.add_error(
+                        f'Failed to process coinbase prime event due to {e}. Skipping entry...',
+                    )
+
+            if (
+                response['pagination']['has_next'] is True and
+                (new_cursor := response['pagination'].get('next_cursor', '')) != ''
+            ):
+                query_params['cursor'] = new_cursor
+            else:
+                break
+
+        return result  # type: ignore  # mypy doesn't detect that the return type here is defined by the function used
```

Comment:
For this don't you think there is a way to do it? Don't want you to spend time here, but some times it's missing some obvious typing thing. Have you tried going in with the `reveal_type()` in various places and seeing what mypy thinks it is?

Reply:
https://github.com/orgs/rotki/projects/11?pane=issue&itemId=85419257 created a note